### PR TITLE
Added Packer and tentative Terraform stages to Jenkins pipeline

### PR DESCRIPTION
**Pull request by duraz0rz**
This adds the Packer build to the Jenkins pipeline. We are also able to grab the AMI ID from the manifest produced by Packer; we echo the ID in a tentative Terraform stage, but the idea is that we can take that ID and pass it to Terraform somehow.

### CODE REVIEW

File: packer/packer.json [hunk 1]
```diff
@@ -1,61 +1,16 @@
...
-    "variables": {
-        "proxy_uname": "{{env `DOGE_PROXY_UNAME`}}",
-        "proxy_password": "{{env `DOGE_PROXY_PASSWORD`}}",
-        "proxy_host": "{{env `DOGE_PROXY_HOST`}}",
-        "proxy_port": "{{env `DOGE_PROXY_PORT`}}",
-        "current_timestamp": "{{timestamp}}"
-    },
     "builders": [
         {
-            "type": "docker",
```
Comment (ztbrown) [inline packer/packer.json:11]: Why are we removing Docker?

File: Jenkinsfile [hunk 2]
```diff
@@ -13,5 +26,20 @@ pipeline {
...
             }
         }
+        stage('Build Application') {
+            steps {
+                sh './gradlew build'
+            }
+        }
+        stage('Build and verify images') {
+            steps {
+                sh './packer/verify'
```
Comment (ztbrown) [inline Jenkinsfile]: Verify is a confusing name for a script that wraps `packer build`.

Comment (ztbrown) [reply Jenkinsfile]: This is going to create a ton of AMIs, right? Do we want to run `packer validate` on branches and PR and `packer build` only on merge to master?

File: Jenkinsfile [hunk 3]
```diff
@@ -13,5 +26,20 @@ pipeline {
...
+            }
+        }
+        stage('Terraform') {
```
Comment (ztbrown) [inline Jenkinsfile]: We want to do this step only on a merge to master.

**Review by ztbrown — COMMENTED** [review]
This could be a topic for our next retro: the history on this branch is confusing (Maisam/Remy commit, merging in master) and I suspect it's because you merged with master instead of rebasing. On my last team, we rebased before pushing to keep linear history. I'm interested in what other people think about this.

File: Jenkinsfile (hunk 2 above)
Comment (d--) [reply Jenkinsfile]: 1.  Yes, it is.  How about we just call it `build`?
2.  Probably a good idea.

Comment (duraz0rz) [reply Jenkinsfile]: `packer validate` doesn't run serverspec. It only checks to see if the syntax and configuration is correct.

**Comment by duraz0rz** [comment]
@ztbrown Maisam and Remy actually did make a commit in our branch to add a stage in the pipeline for us; that wasn't a merge.

How we want to get new commits into our branches (merge vs rebase) can be a debate.

File: packer/packer.json (hunk 1 above)
Comment (d--) [reply packer/packer.json:11]: For speed.  Docker requires us to:
1.  Keep the images in sync.
1.  Set up a proxy.
1.  Add a user.
1.  Account for being root at every step.
1.  Account for running on the dev's machine.
1.  Account for the fact that docker's images are much more minimal than AMI / VM images.
1.  Run docker *in* docker on Jenkins, which is a pain when setting up the Jenkins container.
1.  Figure out why Packer hangs when attempting to build a docker image in a docker image, when doing it manually works fine.

File: Jenkinsfile (hunk 2 above)
Comment (remingtron) [reply Jenkinsfile]: I think ```build``` is certainly an improvement on ```verify```

**Review by remingtron — APPROVED** [review]
Looks good, pending ```verify``` rename

**Pull request merged**